Addition of pixel Tracking in HI pp_on_AA_2018 era

abaty opened the pull request:
This PR creates a new era which is a copy of Run2_2018_pp_on_AA, and adds a pixel tracking sequence to the reconstruction.  This pixel tracking sequence has been requested to meet physics goals requiring low-pt tracks with a higher purity than what is achievable with full tracks.  The additional pixel tracking modules have been found to have a negligible impact on the total timing.  Storing the extra track collection has been estimated to increase the event content size of AOD by 11%, which is a number the HI PAG is willing to accept.  Details of the pixel tracking performance in MB HI events can be found at [1].

The PR can be tested with the normal pp_on_AA_2018 cmsDriver command with the era changed to the 'Run2_2018_pp_on_AA_wPixelTrk'.

[1] https://indico.cern.ch/event/749011/contributions/3098671/attachments/1698408/2736177/pixelUpdate_20180727.pdf

@mandrenguyen

Commit:
added pixelTracking era for pp_on_AA_2018

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24257/5964

cmsbuild commented:
A new Pull Request was created by  @abaty  for master.

It involves the following packages:

Configuration/Eras
Configuration/StandardSequences
RecoHI/HiTracking
RecoTracker/Configuration


@perrotta, @cmsbuild, @franzoni, @slava77, @fabiocos, @davidlange6 can you please review it and eventually sign? Thanks.
@echapon, @makortel, @felicepantaleo, @GiacomoSguazzoni, @rovere, @VinInn, @Martin-Grunewald, @mschrode, @yenjie, @jazzitup, @kurtejung, @gpetruc, @ebrondol, @mandrenguyen, @dgulhan, @yetkinyilmaz this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
> The additional pixel tracking modules have been found to have a negligible impact on the total timing.

Why do we then need a new era?
can this new collection be added to the already existing pp_on_AA_2018 ?

abaty commented:
My understanding is that depending on the final data-taking strategy, there could be a request to only run the pixel tracking on a subset of all the PDs gathered for the HI run in 2018.  In order to facilitate running two different sequences on specific PDs, I was told it would be beneficial to keep the pixel tracking as a separate era.  Let me know if this is not the case. @slava77

slava77 commented:
On 8/9/18 7:03 AM, abaty wrote:
> My understanding is that depending on the final data-taking strategy,
> there could be a request to only run the pixel tracking on a subset of
> all the PDs gathered for the HI run in 2018. In order to facilitate
> running two different sequences on specific PDs, I was told it would be
> beneficial to keep the pixel tracking as a separate era. Let me know if
> this is not the case. @slava77

ok,
from your most recent response it sounds like the statement in the PR
description
"Storing the extra track collection has been estimated to increase the
event content size of AOD by 11%, which is a number the HI PAG is
willing to accept."
is a bit of wishful thinking.

BTW, how does this increment look like for minbias or for something very
central?

mandrenguyen commented:
@slava77 Based on the decision at the last HIN PAG meeting, we are indeed willing to add 11% to AOD.   Regarding the two eras, this was requested for the minbias datasets, so I suggested to make a separate era.   This being said, I don't think it was  yet excluded that we just store pixel tracks for all PDs.   Let us get clarification at tomorrow's PAG meeting and we can update this PR if necessary.

abaty commented:
We recently checked the size increase in b=0 events, and it looks like the number is still 11%.

slava77 commented on RecoHI/HiTracking/python/HILowPtConformalPixelTracks_cfi.py:
```diff
@@ -41,6 +41,89 @@
+        extraHitRPhitolerance = cms.double(0.032),
+        maxChi2 = cms.PSet(
+           enabled = cms.bool(True),
+           pt1 = cms.double(0.7),
```
please drop type specifications for all parameters which already exist.
This is a safer way to protect from parameter name mistakes and will also help in possible parameter migrations.

slava77 commented on RecoHI/HiTracking/python/HILowPtConformalPixelTracks_cfi.py:
```diff
@@ -41,6 +41,89 @@
+)
+
+#Pixel tracks
+hiConformalPixelTracksPhase1 = cms.EDProducer("PixelTrackProducer",
```
I think that "Phase1" can be dropped from the producer name.

slava77 commented:
@mandrenguyen 
please clarify when you expect to have a decision on this.
Thank you.

abaty commented:
Based on the HI PAG meeting on Friday, the decision was made to move the pixel tracking into the main pp_on_AA_2018 era.  To try to mitigate some of the event size increase, we will also slightly bump up the low-pt threshholds of the generalTracks (as these tracks are mostly fake anyways).

I'm finishing up some tests on the new setup and should push is soon.

Commit:
Changed back to keeping pixelTracks in pp_on_AA era

Commit:
removed cms data types

cmsbuild commented:
The code-checks are being triggered in jenkins.

abaty commented:
I've moved the changes in this PR back into the main pp_on_AA_2018 era, as discussed in my previous comment.  The pt threshholds for the low pt steps have been raised from 0.3 to 0.49 GeV as well.  This reduces the number of tracks under 0.5 GeV stored, reducing the extra event content added to AOD in this PR from 11% down to around 7%.  RECO now also runs slightly faster because of the timing saved in the low-pt step.

I've also made changes addressing comments from Slava's code review.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24257/6002

cmsbuild commented:
Pull request #24257 was updated. @perrotta, @cmsbuild, @franzoni, @slava77, @fabiocos, @davidlange6 can you please check and sign again.

slava77 commented:
@cmsbuild please test workflow 158.0

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented:
@abaty 
please edit the PR title and description to match the latest implementation.
There is no new era/modifier.

slava77 commented on RecoHI/HiTracking/python/HILowPtConformalPixelTracks_cfi.py:
```diff
@@ -103,22 +106,22 @@
-    Fitter = cms.InputTag("pixelFitterByConformalMappingAndLine"),
-    SeedingHitSets = cms.InputTag("hiConformalPixelTracksPhase1HitQuadrupletsCA"),
+from Configuration.Eras.Modifier_pp_on_AA_2018_cff import pp_on_AA_2018
+pp_on_AA_2018.toModify(hiConformalPixelTracks,
```
shouldn't this be ```phase1Pixel.toModify```?

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24257/29818/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-24257/158.0_HydjetQ_B12_5020GeV_2018+HydjetQ_B12_5020GeV_2018+DIGIHI2018PPRECO+RECOHI2018PPRECO+HARVESTHI2018PPRECO

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 3 differences found in the comparisons
 - DQMHistoTests: Total files compared: 31
 - DQMHistoTests: Total histograms compared: 2891469
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2891277
 - DQMHistoTests: Total skipped: 190
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 30 files compared)
 - Checked 129 log files, 14 edm output root files, 31 DQM output files

Commit:
changed from pp_on_AA to more general pixelPhase1 era

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24257/6028

cmsbuild commented:
Pull request #24257 was updated. @perrotta, @cmsbuild, @franzoni, @slava77, @fabiocos, @davidlange6 can you please check and sign again.

cmsbuild commented:
+1
Tested at: 6c81153aac8e069a6c6a1bbe320af7b28399eab2
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24257/29838/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24257/29838/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-24257/158.0_HydjetQ_B12_5020GeV_2018+HydjetQ_B12_5020GeV_2018+DIGIHI2018PPRECO+RECOHI2018PPRECO+HARVESTHI2018PPRECO

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 31
 - DQMHistoTests: Total histograms compared: 2891469
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2891278
 - DQMHistoTests: Total skipped: 190
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 30 files compared)
 - Checked 129 log files, 14 edm output root files, 31 DQM output files

slava77 commented:
Here are some observations to go with the PR

Using 230 minbias events after the filter "3000":
there is about 10% drop in the number of generalTracks
![all_sign1034vsorig_hinpbpbmbfilter3000c_log10recotracks_generaltracks__reco_obj_pt](https://user-images.githubusercontent.com/4676718/44365714-76754980-a47f-11e8-9d4b-766cbd1289ef.png)
mostly from low pt.

![all_sign1034vsorig_hinpbpbmbfilter3000c_recotracks_generaltracks__reco_obj_algo](https://user-images.githubusercontent.com/4676718/44366186-cc96bc80-a480-11e8-8398-86050a9d3876.png)
low-pt triplet and quad iterations are down, in line with the increased thresholds, but the later iterations are now picking up more tracks (including pixel pair).

particle flow follows with a drop at very low pt and some increase up to 1.5 GeV, from the later iterations; overall 12% fewer charged candidates
![all_sign1034vsorig_hinpbpbmbfilter3000c_log10recopfcandidates_particleflow__reco_obj_pt46](https://user-images.githubusercontent.com/4676718/44365592-2302fb80-a47f-11e8-82b4-768b892f0623.png)

the pixel tracks show up as expected. There are about 5K tracks per event (only about 20% less than in the generalTracks collection)
![pixeltrack_origvssign1034_hinpbpbmbfilter3000c_log10recotracks_hiconformalpixeltracks__reco_obj_pt](https://user-images.githubusercontent.com/4676718/44366445-8b52dc80-a481-11e8-9bb9-b0a107679ebe.png)


CPU use decreased by about 2%:
- the tracking time total is roughly unchanged : a reduction in the lowPt iterations is balanced out by the added pixel tracks and also increase in time spent on later iterations.
- overall decrease in the number of tracks and PF charged candidates reduces time in various high level reco, which adds up to about 2% and explains a total CPU reduction

The event size is up by about 7.5% in AODSIM output
- hiConformalPixelTracks contribute an increase of about 11%, as expected from the PR description
- the reduction in generalTracks and pf candidates dominates the overall reduction and adds up to 2.3%

Memory peak in a production-like job with 8 threads increased by 1 GB (from 15 GB to 16 GB). This is apparently the cost of adding the pixel tracks, but is perhaps also partly coming from the increase in iterative tracking seeds and candidates for higher iterations.

Overall, the results appear to be somewhat as expected.

slava77 commented:
similar plots for wf 158.0 (200 events):

![all_sign1034vsorig_hydjetqb12pprecoin2018wf158p0c_log10recotracks_generaltracks__reco_obj_pt](https://user-images.githubusercontent.com/4676718/44366946-28fadb80-a483-11e8-9d1b-4c7970296e91.png)

![all_sign1034vsorig_hydjetqb12pprecoin2018wf158p0c_recotracks_generaltracks__reco_obj_algo](https://user-images.githubusercontent.com/4676718/44366988-45971380-a483-11e8-98f0-79318d7dd020.png)

![all_sign1034vsorig_hydjetqb12pprecoin2018wf158p0c_log10recopfcandidates_particleflow__reco_obj_pt46](https://user-images.githubusercontent.com/4676718/44366927-1bddec80-a483-11e8-89e6-1a26964e2454.png)

there is essentially no visible increase in the number of tracks with pt around 1 GeV. Here the fraction of fakes is apparently much smaller.

The change in the fake rate can be seen from the MTV plots:
- for the wf 158.0 it did not increase much and remains low
<img width="740" alt="wf158 0_general09_fr" src="https://user-images.githubusercontent.com/4676718/44367598-1da8af80-a485-11e8-9257-0e9a73a45f1c.png">
- for the "filter 3000" events it has gone up significantly
<img width="739" alt="wffilter3000_general09_fr" src="https://user-images.githubusercontent.com/4676718/44367643-387b2400-a485-11e8-9097-3fa8caab5e04.png">

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/24257	68367e8
- code [configuration] changes are in line with the (updated) PR description and the follow up notes in https://github.com/cms-sw/cmssw/pull/24257#issuecomment-412641054 (the increase in iter tracking thresholds ). The changes affect only the ```pp_on_AA_2018``` modifiers.
- jenkins tests pass and comparisons with the baseline show no differences (the ```pp_on_AA_2018``` is not in the standard tests).
- local tests with wf 158.0 and "filter 3000" events show somewhat expected behavior (see notes in https://github.com/cms-sw/cmssw/pull/24257#issuecomment-414463507 and  https://github.com/cms-sw/cmssw/pull/24257/#issuecomment-414469228)

kpedro88 commented:
+1

kpedro88 commented:
merge
